app/src/main/java/org/jellyfin/androidtv/ui/startup/StartupActivity.kt

```diff
@@ -18,68 +20,67 @@ import org.jellyfin.androidtv.ui.browsing.MainActivity
 import org.koin.android.ext.android.inject
-import org.koin.androidx.fragment.android.replace
 import timber.log.Timber
 
-class StartupActivity : FragmentActivity() {
+class StartupActivity : FragmentActivity(R.layout.fragment_content_view) {
 	companion object {
-		private const val NETWORK_PERMISSION = 1
-		const val ITEM_ID = "ItemId"
-		const val ITEM_IS_USER_VIEW = "ItemIsUserView"
-		const val HIDE_SPLASH = "HideSplash"
+		const val EXTRA_ITEM_ID = "ItemId"
+		const val EXTRA_ITEM_IS_USER_VIEW = "ItemIsUserView"
+		const val EXTRA_HIDE_SPLASH = "HideSplash"
 	}
 
-	private var application: TvApp? = null
 	private val apiClient: ApiClient by inject()
 	private val mediaManager: MediaManager by inject()
 	private val serverRepository: ServerRepository by inject()
 	private val sessionRepository: SessionRepository by inject()
 
+	private val networkPermissionsRequester = registerForActivityResult(
+		ActivityResultContracts.RequestMultiplePermissions()
+	) { grants ->
+		val anyRejected = grants.any { !it.value }
+
+		if (anyRejected) {
+			// Permission denied, exit the app.
+			Toast.makeText(this, R.string.no_network_permissions, Toast.LENGTH_LONG).show()
+			finish()
+		} else {
+			observeSession()
+		}
+	}
+
 	override fun onCreate(savedInstanceState: Bundle?) {
 		super.onCreate(savedInstanceState)
-		setContentView(R.layout.fragment_content_view)
-		if (!intent.getBooleanExtra(HIDE_SPLASH, false)) {
-			supportFragmentManager.beginTransaction()
-				.replace(R.id.content_view, SplashFragment())
-				.commit()
-		}
-		application = applicationContext as TvApp
+
+		if (!intent.getBooleanExtra(EXTRA_HIDE_SPLASH, false)) showSplash()
 
 		// Migrate old credentials
-		lifecycleScope.launch {
+		runBlocking {
 			serverRepository.migrateLegacyCredentials()
 		}
 
 		// Ensure basic permissions
-		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_NETWORK_STATE) != PackageManager.PERMISSION_GRANTED
-				|| ContextCompat.checkSelfPermission(this, Manifest.permission.INTERNET) != PackageManager.PERMISSION_GRANTED)) {
-			Timber.i("Requesting network permissions")
-			ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.ACCESS_NETWORK_STATE, Manifest.permission.INTERNET), NETWORK_PERMISSION)
-		} else {
-			Timber.i("Basic network permissions are granted")
-			start()
-		}
+		networkPermissionsRequester.launch(arrayOf(Manifest.permission.INTERNET, Manifest.permission.ACCESS_NETWORK_STATE))
 	}
 
-	private fun start() {
+	private fun observeSession() {
 		var isLoaded = false
+
 		sessionRepository.currentSession.observe(this) { session ->
 			if (session != null) {
 				Timber.i("Found a session in the session repository, waiting for the currentUser in the application class.")
 
-				supportFragmentManager.beginTransaction()
-					.replace(R.id.content_view, SplashFragment())
-					.commit()
+				showSplash()
```

Comment:
@nielsvanvelzen  Not sure if this `showsplash` is actually necessary, but it breaks the functionality of `EXTRA_HIDE_SPLASH`

Reply:
The extra is used to disable the initial splash screen when starting the activity. If we're loading a session it should still show up so this one is necessary.

Reply:
Obviously a bit nebulous to me ;)
We already have a session because `session != null`, haven't we?

Reply:
Yes but we're still waiting for the apibinder to retrieve user info, which can take a few seconds depending on the network.

Reply:
Sure for both you and me it is pretty much instant, but this is the first api call to the server when autologin is enabled and some people have weird setups that warm up the server on the first connection, we also need to account for server timeouts here, which is set to 6 seconds by default. Not showing anything in this timeframe is bad.